core-rust/core-api-server/src/core_api/handlers/stream_proofs.rs

```diff
@@ -0,0 +1,211 @@
+
+use transaction::prelude::*;
+
+#[tracing::instrument(skip(state))]
+pub(crate) async fn handle_stream_proofs(
+    state: State<CoreApiState>,
+    Json(request): Json<models::StreamProofsRequest>,
+) -> Result<Json<models::StreamProofsResponse>, ResponseError<models::StreamProofsErrorDetails>> {
+    assert_matching_network(&request.network, &state.network)?;
+    let mapping_context = MappingContext::new(&state.network);
+
+    let filter = request
+        .filter
+        .ok_or_else(|| client_error("filter must be present"))?;
+
+    let page_size = extract_valid_size(
+        request.max_page_size,
+        SizeRange {
+            min: 1,
+            default: 20,
+            max: 100,
+        },
+    )
+    .map_err(|err| err.into_response_error("max_page_size"))?;
+
+    let continue_from_state_version =
+        extract_continuation_token::<StateVersion>(request.continuation_token)
+            .map_err(|err| err.into_response_error("continuation_token"))?;
+
+    let database = state.state_manager.database.read_current();
+
+    use models::StreamProofsFilter::*;
+    let mut proofs_iter = match filter {
+        StreamProofsFilterAll { from_state_version } => iterate_all_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_state_version(&database, from_state_version)?,
+        ),
+        StreamProofsFilterNewEpochs { from_epoch } => iterate_end_of_epoch_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_epoch(&mapping_context, &database, from_epoch)?,
+        ),
+        StreamProofsFilterProtocolUpdateInitializations { from_state_version } => {
+            iterate_protocol_update_initialization_proofs(
+                &database,
+                continue_from_state_version,
+                extract_from_state_version(&database, from_state_version)?,
+            )
+        }
+        StreamProofsFilterProtocolUpdateExecution {
+            from_state_version,
+            protocol_version,
+        } => iterate_protocol_update_execution_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_state_version(&database, from_state_version)?,
+            protocol_version,
+        ),
+    }?;
+
+    let (page, continuation_token) = to_api_page(
+        &mut proofs_iter,
+        page_size,
+        |proof| handlers::to_api_ledger_proof(&mapping_context, proof),
+        |proof| proof.ledger_header.state_version,
+    )?;
+
+    Ok(Json(models::StreamProofsResponse {
+        page,
+        continuation_token,
+    }))
+}
+
+fn iterate_all_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_state_version: StateVersion,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let start = optional_max(from_state_version, continue_from_state_version);
```

Comment:
couldn't we just disregard the `from_state_version` when continuation token is present?

(for bonus points: `if from_state_version > continue_from_version { return Err("you tried to use the continuation token with different filters!") }`)

Reply:
True, we could. I honestly don't think it matters. We agreed that behaviour is undefined if the user changes their filters.

I think this feels the most natural way of handling such a case, but I don't mind too much.

Reply:
> We agreed that behaviour is undefined if the user changes their filters.

hah, `continuation_token.or_else(from_state_version)` is "undefined behavior" too... it is a shorter impl, and ofc can confuse user in a different way than `optional_max()`.

TBH the more I think, the more I am convinced that `HTTP 400: please don't tamper with continuation token and filters` is worth it (and AFAIU, Krzysztof too). But that's a topic for a future API conventions adjustment.